# Pokemón

## Tarea 3:
## Juego de cartas pokemón, combinaciones de mano

In [55]:
library(prob)

Pokémon: Juego de Cartas Coleccionables es un juego de cartas coleccionables competitivo para dos jugadores basado en la saga de videojuegos Pokémon.


En una partida de Pokémon Trading Card Game, los jugadores asumen el papel de un entrenador Pokémon durante un combate de forma análoga a los videojuegos. Para poder jugar, cada jugador debe tener una baraja de 60 cartas que habrá preparado previamente. La composición de las barajas es libre.

**Fuente:**

[Pokémon Trading Card Game](https://es.wikipedia.org/wiki/Pokémon_Trading_Card_Game)

Tomando en cuenta que existen 801 pokemones en nuestra base de datos ¿Cuantas posibles manos existen dentro del juego Pokemón?

In [19]:
#se escogen 60 cartas de un total de 801
# se trata de una combinación 801 C  60

choose(801, 60)

[1] 2.057847e+91

In [20]:
cat('Existe un total de:', choose(801, 60) , 'distintas manos dentro del juego')

Existe un total de: 2.057847e+91 distintas manos dentro del juego

**Existen practicamente una infinidad de posibles manos al jugar este juego.**

## Inicio de Juego

"En primer lugar, los jugadores deberán barajar sus respectivas barajas. A continuación, cada uno robará 7 cartas que se quedará en su mano sin mostrarlas al contrincante. Si alguno de los dos jugadores no posee en su mano al menos un Pokémon básico (es decir, un Pokémon que no sea evolución de otro), entonces deberá mostrar sus cartas al oponente para demostrarlo; devolver sus cartas, barajar de nuevo, y robar 7 cartas. El oponente, a cambio, puede robar una carta adicional. El proceso se debe de hacer hasta que ambos jugadores posean al menos en su mano un Pokémon básico." (wikipedia)

In [2]:
#se escogen 7 cartas de un total de 60
# se trata de una combinación 60 C  7
choose(60, 7)

[1] 386206920

**Con las 60 cartas de incio se pueden tener 386,206,920 formas distintas de iniciar una partida**

Si la regla de inicio dice que debes de tener al menos un Pokémon básico, ¿cuál es la probabilidad de escogerlo entre las 7 primeras cartas de inicio si solo tienes uno de ellos en tus 60 cartas?

**devido a que sacar 7 cartas en sucesión y sin reemplazo no son eventos independientes la probabilidad de sacar una determinada carta se puede calcular con la siguente expresion**

$$ P(S) = P(\{e_1, e_2, \dots , e_n\}) = P(e_1 \cup e_2 \cup \dots \cup e_n) = \displaystyle \sum_{i=1}^n P(e_i)$$

para este caso seria:

$$\displaystyle \sum_{i=1}^k \frac{1}{n-i+1}$$


In [26]:
#parametros 
# c = numero de cartas iguales en la baraja
# n = numero de cartas en la baraja
# k = numero de cartas a extraer en una mano
cal_prob_cartas <- function(c,n,k){
    res <- 0
    for (i in 1:k) {
    res <- res + c/(n-i+1)
        }
     return(res)
}


[1] 0.1229585

**si solo Sí solo se tiene una carta de Pokémon básico, la probabilidad de sacarla en un mazo de 60 cartas, eligiendo 7 al azar es:**

In [28]:
cal_prob_cartas(1,60,7)


[1] 0.1229585

Dado que hay una penalización si no se obtiene una carta de Pokémon básico, la probabilidad obtenida 0.12 al tener solo una en el mazo es muy baja. Por lo tanto se desea saber cuántas cartas Pokémon básico se necesitan para tener una probabilidad cercana a uno.

Para este caso se puede calcular como:

$$C =\frac{P}{\displaystyle \sum_{i=1}^k \frac{1}{n-i+1}} $$

In [32]:
#parametros 
# c = numero de cartas iguales en la baraja
# n = numero de cartas en la baraja
# k = numero de cartas a extraer en una mano
# p = probabilidad deseada
cal_num_cartas_iguales <- function(p,n,k){
    aux <- 0
    for (i in 1:k) {
    aux <- aux + 1/(n-i+1)
        }
    c <- p/aux
     return(c)
}

usando la formula para tener una probalidad .9 o mayor es

In [53]:
cal_num_cartas_iguales(.9,60,7)

[1] 7.319541

**por lo tanto si se tienen 8 cartas de Pokeon Basicos, se tiene una probilidad mayor a .9 de sacar un Pokeon Basico en la primera mano**